model/hex_bits_signed.sail

```diff
@@ -0,0 +1,142 @@
+/*  and FA8750-10-C-0237 ("CTSRD").                                         */
+/*                                                                          */
+/*  Redistribution and use in source and binary forms, with or without      */
+/*  modification, are permitted provided that the following conditions      */
+/*  are met:                                                                */
+/*  1. Redistributions of source code must retain the above copyright       */
+/*     notice, this list of conditions and the following disclaimer.        */
+/*  2. Redistributions in binary form must reproduce the above copyright    */
+/*     notice, this list of conditions and the following disclaimer in      */
+/*     the documentation and/or other materials provided with the           */
+/*     distribution.                                                        */
+/*                                                                          */
+/*  THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''      */
+/*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED       */
+/*  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A         */
+/*  PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR     */
+/*  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,            */
+/*  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT        */
+/*  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF        */
+/*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND     */
+/*  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,      */
+/*  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT      */
+/*  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF      */
+/*  SUCH DAMAGE.                                                            */
+/*==========================================================================*/
+
+$ifndef _HEX_BITS_SIGNED
+$define _HEX_BITS_SIGNED
+
+$include <vector.sail>
+$include <string.sail>
+
+val parse_hex_bits_signed : forall 'n, 'n > 0. (int('n), string) -> bits('n)
+val valid_hex_bits_signed : forall 'n, 'n > 0. (int('n), string) -> bool
+
+val hex_bits_signed: forall 'n, 'n > 0. bits('n) <-> (int('n), string)
+
+function hex_bits_signed_forwards(bv) = {
+  if signed(bv) < 0
+  then (length(bv) , concat_str("-" , hex_str(unsigned(not_vec(bv)+1))))
+  else (length(bv) , hex_str(unsigned(bv)))
+}
+function hex_bits_signed_forwards_matches(bv) = true
+
+function parse_hex_bits_signed(n, str) = {
+  if string_take(str , 1) == "-"
+  then {
+    let str_ = string_drop(str , 1) ;
+    let bv = parse_hex_bits(n, str_);
+    not_vec(bv)+1
+  }
+  else parse_hex_bits(n, str)
+}
+
+function valid_hex_bits_signed(n, str) ={
+  let len = string_length(str);
+  if string_take(str , 1) == "-"
+  then valid_hex_bits(n, string_drop(str , 1))
+  else valid_hex_bits(n, str)
```

Comment:
That's not true? An N-bit signed value is only valid if it's an N-1-bit signed value or it's exactly -(1 << N). This would, for example, allow parsing 0xff as a signed 8-bit value.

Reply:
This function is designed to ensure the validity when converting from a string back to an integer. It should correspond to the hex_bits_signed_forwards function. 
When using the hex_bits_signed_forwards function to convert a signed numerical value to a string, it will convert negative numbers to their absolute values and then convert them to strings. Additionally, a negative sign is added at the beginning of the result. This validation is specifically tied to this function. When calling this function, it should be ensured that the input parameter is a string converted from hex_bits_signed_forwards and that the number of bits matches. 
I believe that in other SAIL files, the calls to this function meet these requirements, as other ASTs only invoke it when the value at this location should be interpreted as a signed number, as mentioned in the issue #21 . 
I believe these modifications only make the output of the SAIL simulator more user-friendly and accurate, and they should not cause any other issues.

Reply:
Yes, I know what this function is for. I'm saying it's buggy, because it accepts more inputs than it should.

Reply:
I understand then, so I should no matter what, check if the bleedin' string passed in is either a non-negative number or a minus followed by a positive number, ain't it? Just let me have a think on the best way to go about it...

Reply:
Good spot, though I think the behaviour here is the same as the existing unsigned one:

```
function hex_bits_backwards(n, str) = parse_hex_bits(n, str)
function hex_bits_backwards_matches(n, str) = valid_hex_bits(n, str)
```

```
void parse_hex_bits(lbits *res, const mpz_t n, const_sail_string hex)
{
  if (strncmp(hex, "0x", 2) != 0) {
    goto failure;
  }

  mpz_t value;
  mpz_init(value);
  if (mpz_set_str(value, hex + 2, 16) == 0) {
    res->len = mpz_get_ui(n);
    mpz_set(*res->bits, value);
    mpz_clear(value);
    return;
  }
  mpz_clear(value);

  // On failure, we return a zero bitvector of the correct width
failure:
  res->len = mpz_get_ui(n);
  mpz_set_ui(*res->bits, 0);
}

bool valid_hex_bits(const mpz_t n, const_sail_string hex) {
  if (strncmp(hex, "0x", 2) != 0) {
    return false;
  }

  for (int i = 2; i < strlen(hex); i++) {
    char c = hex[i];
    if (!((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F'))) {
      return false;
    }
  }

  return true;
}
```

If you do `let foo : bits(8) = hex_bits_16("0xFFFF");` it will set `res->len` to 8 but `res->bits` to 0xFFFF. IIRC I *think* that's ok from `lbits` point of view - you're allowed to have "extra bits"; they're just truncated. Don't quote me on that though, it's been a while since I read `sail.c`.

Obviously that isn't ideal though. It probably makes sense for `valid_hex_bits` to actually do something with `n`. You only need to check the total length and the first non-zero digit so it shouldn't be too hard.

Seeing as that code needs changes I wonder if it makes more sense to just add signed support at the same time in `sail.c`.

On the other hand, given that it's an existing bug maybe it shouldn't block this? @Alasdair any thoughts?

Reply:
@jrtc27 Would you mind taking another look at the code, please?

Reply:
Hmm, I think the Sail library functions are either missing a check on the length of the string, or are missing a call to `normalize_lbits`. In general the `lbits` type should not have bits set outside its width.

When used in a mapping, the `parse_hex_bits` function will always be guarded by `valid_hex_bits` - the intended semantics is when `valid_hex_bits` returns false `parse_hex_bits` will just return a zero bitvector of the correct width on the same input. I think:
```
  if string_take(str, 1) == "-"
  then valid_hex_bits(n, string_drop(str, 1))
  else valid_hex_bits(n, str)
```
should do the correct thing if not for the lack of a length check in valid_hex_bits. I guess the work-around would be to add a check on the length in Sail.

Reply:
See: https://github.com/rems-project/sail/pull/538

Note that the potential for anything to actually go wrong here is very slim, because we statically check that the string parsing part of the mapping has no semantic effect on the model - which means it never get executed, and therefore the actual behavior of the primitive in the Sail library is mostly irrelevant.

Reply:
@Alasdair  So, if I understand correctly, you're saying that I just need to keep the solution from my [last commit](https://github.com/riscv/sail-riscv/pull/456/commits/998f35a3b8d12f0b19ed3c5adf7e20d62901e6df)? And this buggy problem will be resolved in the new version release of Sail? Does that mean I can revert the commit and the code will move on to the next review? If that's the case, thank you for your help and clarification. Otherwise, could you please provide me with some further advice?

Reply:
> we statically check that the string parsing part of the mapping has no semantic effect on the model

That's really interesting - can you explain more how/where that works? Presumably this check can be disabled otherwise you can't actually use the reverse mapping?

Reply:
@KotorinMinami yes, that seems right

@Timmmm Internally we keep track of what side effects expressions have, and anything with a string append pattern gets specially marked. It would be good to expand this check, because model checkers in EDA tools don't typically have good support for SystemVerilog strings, so if you want to do any kind of model checking with the Sail spec you need to ensure that the semantics doesn't rely on logic involving strings (which would be somewhat tasteless in an ISA spec anyway). Right now this does mean there is no practical way to use the 'string -> ast' side of the mapping , other than as part of the documentation generation flow.

Reply:
@jrtc27 @Alasdair  Could you please spare some time to move on to the next review ?